<a href="https://colab.research.google.com/github/akechsmith/ICS-2207-SCIENTIFIC-COMPUTING/blob/main/draft_h5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

NAME: AKECH DAU ATEM

REG NO: SCT211-0535/2022

COURSE: BSC COMPUTER SCIENCE

UNIT: ICS2207 SCIENTIFIC COMPUTING

PROJECT TITLE: APPLICATION OF NEWTON-RAPHSON METHOD IN FINDING ROOTS OF EQUATIONS


DEADLINE: MARCH 6, 2024



---




OBJECTIVE:

Utilize the Newton-Raphson method to numerically solve for the roots of a given
set of equations. The project will assess your ability to apply this iterative method effectively,
analyze convergence criteria, and interpret the results in the context of scientific computing

1. INTRODUCTION


The Newton-Raphson method is named after Isaac Newton; the man who discovered the method in 1736, and Joseph Raphson, the man who described the method back in 1690. Both mathematicians utilized calculus in this method in order to find the roots of an equation

The Newton-Raphson method, a fundamental technique in numerical analysis, plays a pivotal role in approximating roots of real-valued functions. This iterative approach provides an efficient means of finding solutions to equations where explicit solutions may be challenging or impractical to obtain.

Central to its significance is the method's ability to converge rapidly towards the root of a function. By iteratively refining an initial guess, the Newton-Raphson method demonstrates remarkable efficiency in solving nonlinear equations. This method is particularly valuable in diverse fields such as physics, engineering, and finance, where numerical approximations are frequently employed to solve complex problems.

In this project, we explore the application of the Newton-Raphson method to solve nonlinear equations, assessing its efficacy and significance in the broader context of scientific computation.

2. THEORY

## Formula ##
The Newton-Raphson method is an iterative numerical technique used for finding successively better approximations to the roots (or zeros) of a real-valued function. Given a function $f(x)$ and its derivative $f'(x)$,The Newton-Raphson iteration formula is given by:  $ \;x_{n+1}$ = $x_{n}$ - $ \frac{f(x_n)}{f'(x_n)} $




Here:

$ \;x_{n+1}$ is the next approximation,
$\;x_{n}$ is the current approximation, $\;f(x_n)$ is the function value at $\;x_{n}$
$\;f'(x_n)$is the derivative of the function at $\;x_{n}$
## Convergence Criteria ##

For the Newton-Raphson method to converge,$\;x_{0}$ the initial guess
should be close to the actual root, and the function
$f(x)$ must be continuous and have a continuous derivative in the neighborhood of the root. The convergence criteria can be stated using the error term:

|$ \;x_{n+1}$ -  $x_{n}$|  $\; \le$ $\;ϵ$

where
ϵ is a small positive number, indicating the desired level of accuracy. Another common criterion is to check that  |$f(\;x_{n+1})$ |  $\; \le$ $\;ϵ$  
The convergence rate of the Newton-Raphson method is quadratic, which means that the number of correct digits roughly doubles with each iteration when it is converging.


## Potential Limitations ##
While the Newton-Raphson method is a powerful tool for finding roots, it is important to consider its limitations and ensure that the chosen initial guess and function properties align with the method's requirements for convergence:

### Dependency on Initial Guess:
The Newton-Raphson method is sensitive to the choice of the initial guess. If the initial guess is far from the actual root or is in a region where the function has a flat slope, the method may fail to converge.

### Convergence to Local Minimum:
The method is not guaranteed to converge to the global minimum; it might converge to a local minimum or a saddle point.

### Discontinuity and Singularities:
If the function has discontinuities or singularities in the vicinity of the root, the method may behave unpredictably.

### Need for Derivative:
The method requires the calculation of the derivative, which may not always be readily available or computationally expensive.

### Divergence:
The Newton-Raphson method can diverge, especially if the derivative is close to zero or if the iterations lead to a division by zero.

### Not Suitable for Multiple Roots:
If the function has multiple roots, the method may converge to any of them, and the choice of the initial guess becomes crucial.

3. IMPLEMENTATION

Newton-Raphson method in Python to solve a specific equation or set of equations.

exapmle case: $\:f(x)$ = $x^4 - x - 10$


In [ ]:
# Python implementation of the Newton-Raphson method

def newton_raphson_method(f, df, x0, tolerance=1e-6, max_iterations=100):
    """
    Newton-Raphson method for finding the root of a function.

    Parameters:
    - f: The target function.
    - df: The derivative of the target function.
    - x0: Initial guess.
    - tolerance: Convergence tolerance.
    - max_iterations: Maximum number of iterations.

    Returns:
    - root: Approximated root.
    - iterations: Number of iterations performed.
    """

    x = x0
    iterations = 0

    while abs(f(x)) > tolerance and iterations < max_iterations:
        x = x - f(x) / df(x)
        iterations += 1

    return x, iterations

# Example usage
# result = newton_raphson_method(...)


Explanation:

* The newton_raphson_method function takes four parameters: the target function **f**, its derivative **df**, the initial guess **x0**, and **optional parameters** for convergence tolerance and maximum iterations.
* It initializes the current approximation x to the initial guess and
the iteration counter iterations to zero.
* The method iteratively updates the approximation using the Newton-Raphson formula until either the absolute value of the function is below the tolerance or the maximum number of iterations is reached.

This implementation sets the stage for solving a specific equation using the Newton-Raphson method. The next step would be to apply this function to the equation $\:f(x)$ = $x^4 - x - 10$

4. RESULTS AND ANALYSIS

*Example Equation:* $\:f(x)$ = $x^4 - x - 10$

In [ ]:
# Applying Newton-Raphson to the new example equation
f = lambda x: x**4 - x - 10
df = lambda x: 4*x**3 - 1

initial_guess = 2
result, num_iterations = newton_raphson_method(f, df, initial_guess)

# Printing results
print(f"Root found: {result:.3f}")  #root in 3 dp
print(f"Iterations: {num_iterations}")


Root found: 1.856
Iterations: 3


### Results
The Newton-Raphson method is applied to the equation.
the with an initial guess of
$x_0\;$= 2.  The result is a root found at approximately
**1.856** with three decimal places, and the number of iterations required for convergence is **3**.

The method was applied with various intial guesses also. the number of iterations required for convergence for intial guesses 1 and 3 were 8 and 5 respectively.


### Efficiency and Convergence
Analyzing the efficiency involves considering the number of iterations required for convergence. In this case:

The method exhibits high efficiency when the initial guess is
2, converging in only 3 iterations.
For an initial guess of
1, the method still converges but with
8 iterations, suggesting a slightly slower convergence.
Similarly, with an initial guess of
3, the method converges in
5 iterations.
These results indicate that the efficiency of the method is influenced by the choice of the initial guess. A more informed initial guess results in faster convergence.
### Accuracy
To assess accuracy, we compare the obtained root (
1.856) with any known solution. While the exact root of the equation may not be known, accuracy is evaluated based on the consistent convergence to the value
1.856 with different initial guesses. This indicates the accuracy and reliability of the Newton-Raphson method for this equation.

### Challenges Encountered During Implementation
Challenges encountered during the implementation include sensitivity to the initial guess. The number of iterations is significantly affected by the choice of the initial guess, as observed in the variations between
3, 8, and 5 iterations for different initial guesses.



5. CONCLUSION

In summary, the application of the Newton-Raphson method to the nonlinear equation has provided valuable insights into its behavior and performance. The efficiency and accuracy of the method were evident, particularly when utilizing well-informed initial guesses. The impact of the initial guess on convergence was highlighted, showcasing the method's sensitivity to this crucial parameter.

The Newton-Raphson method, with its rapid convergence under favorable conditions, stands as a robust tool for root-finding in scientific computing. Its application, however, requires careful consideration of the initial guess, as demonstrated by variations in the number of iterations for different starting points. This sensitivity emphasizes the importance of understanding the characteristics of the target function.

In broader scientific and computational contexts, the Newton-Raphson method proves its worth, offering an effective balance between efficiency and accuracy. Its limitations, such as sensitivity to poorly-behaved functions or inadequate initial guesses, should be acknowledged. Practitioners may consider alternative methods in scenarios where these limitations become significant.

As a versatile numerical technique, the Newton-Raphson method remains a valuable asset in the toolkit of computational methods, especially when dealing with well-behaved functions and informed initial guesses. Its successful application underscores the significance of thoughtful parameter selection and an understanding of the mathematical properties of the functions being analyzed.

6. REFERENCES

      

1.   Robertson, E. F., and J. J. O'Connor."Joseph Raphson". MacTutor History of Mathematics Archive. Accessed January 13, 2016. http://www-groups.dcs.st-and.ac.uk/~history/Biographies/Raphson.html.

2.    Shodor, (1996). Title of the second source. "Newton-Raphson Method.".  UNC-Chapel Hill Chemistry Fundamentals
    Program.